### PR TITLE
feat!: switch the proposing sequencer after unbonding

### PR DESCRIPTION
**Pull request by mtsitrin**
## Description

----

Closes #XXX

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/sequencer/keeper/sequencer.go [hunk 1]
```diff
@@ -37,10 +37,22 @@ func (k Keeper) UpdateSequencer(ctx sdk.Context, sequencer types.Sequencer, oldS
 	seqByrollappKey := types.SequencerByRollappByStatusKey(sequencer.RollappId, sequencer.SequencerAddress, sequencer.Status)
 	store.Set(seqByrollappKey, b)
 
-	//handle unbonding queue changes due to status change
 	if sequencer.Status == oldStatus {
 		return
 	}
+
+	//rotate proposer if needed
+	if oldStatus == types.Proposer {
```
Comment (omritoptix) [inline x/sequencer/keeper/sequencer.go:45]: dont think it's the cleanest to have this recursive logic here as this function should update the sequencer and it's doing more like rotating.
I'd keep this function clean and just update sequencer status and other implications of updating the sequencer 
 (like rotation etc) would put outside this function.

Comment (mtsitrin) [reply x/sequencer/keeper/sequencer.go:45]: gonna refactor

**Comment by omritoptix** [comment]
@mtsitrin tests failed

File: x/sequencer/keeper/invariants.go [hunk 2]
```diff
@@ -85,23 +85,21 @@ func SequencersPerRollappInvariant(k Keeper) sdk.Invariant {
...
 					}
+					proposerFound = true
 				}
-				if !proposerFound {
-					broken = true
-					msg += "no proposer for rollapp " + rollapp.RollappId + "\n"
-				}
-			*/
+			}
+			if !proposerFound {
```
Comment (omritoptix) [inline x/sequencer/keeper/invariants.go:98]: not sure it's a invariant break. we can have a frozen rollapp or a rollapp with no active sequencer which is waiting for a sequencer to come up. we can't force a rollapp to have a proposer - e.g if a proposer is unbonding and don't want to propose anymore or it's jailed.

File: x/sequencer/keeper/sequencer.go [hunk 3]
```diff
@@ -31,21 +31,27 @@ func (k Keeper) UpdateSequencer(ctx sdk.Context, sequencer types.Sequencer, oldS
...
-		k.setUnbondingSequencerQueue(ctx, sequencer)
-	} else if oldStatus == types.Unbonding {
-		k.removeUnbondingSequencer(ctx, sequencer)
+}
+
+// rotate proposer. we set the first bonded sequencer to be the proposer
+func (k Keeper) RotateProposer(ctx sdk.Context, rollappId string) {
+	seqsByRollapp := k.GetSequencersByRollappByStatus(ctx, rollappId, types.Bonded)
+	if len(seqsByRollapp) == 0 {
+		k.Logger(ctx).Error("no bonded sequencer found for rollapp", "rollappId", rollappId)
```
Comment (omritoptix) [inline x/sequencer/keeper/sequencer.go:48]: not sure it's an error. as I see it's a valid world state.

File: x/sequencer/keeper/invariants.go (hunk 2 above)
Comment (mtsitrin) [reply x/sequencer/keeper/invariants.go:98]: we have 
```
if len(bonded) == 0 {
				continue
			}
			
			
So if there are bonded sequencers, i would expect a proposer one

File: x/sequencer/keeper/sequencer.go (hunk 3 above)
Comment (mtsitrin) [reply x/sequencer/keeper/sequencer.go:48]: right. will change in upcoming PR

**Review by omritoptix — APPROVED** [review]

**Pull request merged**